In [1]:
import pandas as pd

In [2]:
#Reading the Dataset
df=pd.read_csv(r"Data Sets\scraped_novel_dataset.csv")

In [3]:
#Creating a temporary  dataset to work on
novel=df.copy()

In [4]:
novel

Novel_Name     Series_Name  \
0     Clifford's War: The Bluegrass Battleground             NaN   
1                                 Jurassic Park    Jurassic Park   
2                                      Sunburst   Sky King Ranch   
3                            The Paris Apartment             NaN   
4                                       Dempsey         Tier One   
...                                          ...             ...   
7035                                 They Thirst             NaN   
7036                                        Snap             NaN   
7037                  The Traveling Vampire Show             NaN   
7038     WISHBONE...Be Careful What You Wish For             NaN   
7039                            Dropcloth Angels             NaN   

              Author_Name              Genre  Average_Rating No_of_Ratings  \
0        J. Denison Reed    Action-Adventure            4.42            33   
1        Michael Crichton   Action-Adventure            4.09      9,44,442   
2       Susan May Warren    Action-Adventure            4.43         1,467   
3             Lucy Foley    Action-Adventure            3.67      4,73,386   
4         Brian  Andrews    Action-Adventure            4.58         1,708   
...                   ...                ...             ...           ...   
7035      Robert McCammon  Horror/Paranormal            3.91        12,688   
7036      Steven Bredice   Horror/Paranormal            3.40            43   
7037       Richard Laymon  Horror/Paranormal            3.83         8,813   
7038     Brooklyn Hudson   Horror/Paranormal            3.73           764   
7039  Gerald D. Johnston   Horror/Paranormal            4.27            15   

     No_of_Reviews   Score No_of_Votes       Publish_Year  Pages  
0               16     398           4       May 27, 2021  280.0  
1           19,172  30,359         307   November 7, 1990  466.0  
2              335     191           2       June 1, 2022  368.0  
3           40,240     296           3  February 22, 2022  360.0  
4               86   1,260          13  February 21, 2023  350.0  
...            ...     ...         ...                ...    ...  
7035           564   3,615          46    January 1, 1981  412.0  
7036            11     200           2    January 1, 2012  200.0  
7037           645   3,577          40      April 1, 2000  400.0  
7038            94     200           2  February 20, 2012  365.0  
7039             4     200           2   October 29, 2012  480.0  

[7040 rows x 11 columns]

In [5]:
novel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7040 entries, 0 to 7039
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Novel_Name      7040 non-null   object 
 1   Series_Name     4145 non-null   object 
 2   Author_Name     7040 non-null   object 
 3   Genre           7040 non-null   object 
 4   Average_Rating  7040 non-null   float64
 5   No_of_Ratings   7040 non-null   object 
 6   No_of_Reviews   7017 non-null   object 
 7   Score           7040 non-null   object 
 8   No_of_Votes     7040 non-null   object 
 9   Publish_Year    7002 non-null   object 
 10  Pages           6891 non-null   float64
dtypes: float64(2), object(9)
memory usage: 605.1+ KB


### Dropping and replacing

In [6]:
#Replacing the null values of "Series_Name" column with '-' since it is an object type.
novel["Series_Name"].fillna('-',inplace=True)

In [7]:
# dropping the "Pages" column since there are 149 missing values and a lot of data are incorrecct.
novel.drop(["Pages"],axis=1,inplace=True)

In [8]:
novel.isna().sum()

Novel_Name         0
Series_Name        0
Author_Name        0
Genre              0
Average_Rating     0
No_of_Ratings      0
No_of_Reviews     23
Score              0
No_of_Votes        0
Publish_Year      38
dtype: int64

In [9]:
#Checking the number and percentage of null values 
print(novel.isna().sum().sum())
print(novel.isna().sum().sum()/novel.shape[0]*100)

61
0.8664772727272727


In [10]:
#Dropping the null values since there are less than 1% null values in the dataset
novel.dropna(axis=0,inplace=True)
novel=novel.reset_index(drop=True)
novel

Novel_Name     Series_Name  \
0     Clifford's War: The Bluegrass Battleground               -   
1                                 Jurassic Park    Jurassic Park   
2                                      Sunburst   Sky King Ranch   
3                            The Paris Apartment               -   
4                                       Dempsey         Tier One   
...                                          ...             ...   
6997                                 They Thirst               -   
6998                                        Snap               -   
6999                  The Traveling Vampire Show               -   
7000     WISHBONE...Be Careful What You Wish For               -   
7001                            Dropcloth Angels               -   

              Author_Name              Genre  Average_Rating No_of_Ratings  \
0        J. Denison Reed    Action-Adventure            4.42            33   
1        Michael Crichton   Action-Adventure            4.09      9,44,442   
2       Susan May Warren    Action-Adventure            4.43         1,467   
3             Lucy Foley    Action-Adventure            3.67      4,73,386   
4         Brian  Andrews    Action-Adventure            4.58         1,708   
...                   ...                ...             ...           ...   
6997      Robert McCammon  Horror/Paranormal            3.91        12,688   
6998      Steven Bredice   Horror/Paranormal            3.40            43   
6999       Richard Laymon  Horror/Paranormal            3.83         8,813   
7000     Brooklyn Hudson   Horror/Paranormal            3.73           764   
7001  Gerald D. Johnston   Horror/Paranormal            4.27            15   

     No_of_Reviews   Score No_of_Votes       Publish_Year  
0               16     398           4       May 27, 2021  
1           19,172  30,359         307   November 7, 1990  
2              335     191           2       June 1, 2022  
3           40,240     296           3  February 22, 2022  
4               86   1,260          13  February 21, 2023  
...            ...     ...         ...                ...  
6997           564   3,615          46    January 1, 1981  
6998            11     200           2    January 1, 2012  
6999           645   3,577          40      April 1, 2000  
7000            94     200           2  February 20, 2012  
7001             4     200           2   October 29, 2012  

[7002 rows x 10 columns]

### Removing the comma and  changing the datatypes.

In [11]:
novel["No_of_Ratings"]=novel["No_of_Ratings"].str.replace(',','').astype('int')

In [12]:
novel["No_of_Reviews"]=novel["No_of_Reviews"].str.replace(',','').astype('int')

In [13]:
novel["No_of_Votes"]=novel["No_of_Votes"].str.replace(',','').astype('int')

In [14]:
novel["Score"]=novel["Score"].str.replace(',','').astype('int')

In [15]:
novel

Novel_Name     Series_Name  \
0     Clifford's War: The Bluegrass Battleground               -   
1                                 Jurassic Park    Jurassic Park   
2                                      Sunburst   Sky King Ranch   
3                            The Paris Apartment               -   
4                                       Dempsey         Tier One   
...                                          ...             ...   
6997                                 They Thirst               -   
6998                                        Snap               -   
6999                  The Traveling Vampire Show               -   
7000     WISHBONE...Be Careful What You Wish For               -   
7001                            Dropcloth Angels               -   

              Author_Name              Genre  Average_Rating  No_of_Ratings  \
0        J. Denison Reed    Action-Adventure            4.42             33   
1        Michael Crichton   Action-Adventure            4.09         944442   
2       Susan May Warren    Action-Adventure            4.43           1467   
3             Lucy Foley    Action-Adventure            3.67         473386   
4         Brian  Andrews    Action-Adventure            4.58           1708   
...                   ...                ...             ...            ...   
6997      Robert McCammon  Horror/Paranormal            3.91          12688   
6998      Steven Bredice   Horror/Paranormal            3.40             43   
6999       Richard Laymon  Horror/Paranormal            3.83           8813   
7000     Brooklyn Hudson   Horror/Paranormal            3.73            764   
7001  Gerald D. Johnston   Horror/Paranormal            4.27             15   

      No_of_Reviews  Score  No_of_Votes       Publish_Year  
0                16    398            4       May 27, 2021  
1             19172  30359          307   November 7, 1990  
2               335    191            2       June 1, 2022  
3             40240    296            3  February 22, 2022  
4                86   1260           13  February 21, 2023  
...             ...    ...          ...                ...  
6997            564   3615           46    January 1, 1981  
6998             11    200            2    January 1, 2012  
6999            645   3577           40      April 1, 2000  
7000             94    200            2  February 20, 2012  
7001              4    200            2   October 29, 2012  

[7002 rows x 10 columns]

In [16]:
novel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7002 entries, 0 to 7001
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Novel_Name      7002 non-null   object 
 1   Series_Name     7002 non-null   object 
 2   Author_Name     7002 non-null   object 
 3   Genre           7002 non-null   object 
 4   Average_Rating  7002 non-null   float64
 5   No_of_Ratings   7002 non-null   int32  
 6   No_of_Reviews   7002 non-null   int32  
 7   Score           7002 non-null   int32  
 8   No_of_Votes     7002 non-null   int32  
 9   Publish_Year    7002 non-null   object 
dtypes: float64(1), int32(4), object(5)
memory usage: 437.8+ KB


There are 7002 rows and 10 columns

### Formatting the 'Publish_Year' and changing its datatype

In [17]:
import calendar
month_dict = dict((v, k) for k, v in enumerate(calendar.month_name))

In [18]:
month_change=lambda x: month_dict[x]

In [19]:
novel[["month","date","year"]]=novel["Publish_Year"].str.split(' ',expand=True)

In [20]:
novel["date"]=novel["date"].str.replace(',','')

In [21]:
novel["month"]=novel["month"].apply(month_change).astype("str")

In [22]:
novel["Publish_Year"]=novel["date"]+'-'+novel["month"]+'-'+novel["year"]

In [23]:
novel.drop(novel.loc[:,'month':'year'],axis=1,inplace=True)

In [24]:
#changes the datatype to datetime and gives null values to the datas which are not convertible 
novel["Publish_Year"]=pd.to_datetime(novel["Publish_Year"],errors='coerce')
novel

C:\Users\dawap\AppData\Local\Temp\ipykernel_11056\3639606392.py:2: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  novel["Publish_Year"]=pd.to_datetime(novel["Publish_Year"],errors='coerce')


Novel_Name     Series_Name  \
0     Clifford's War: The Bluegrass Battleground               -   
1                                 Jurassic Park    Jurassic Park   
2                                      Sunburst   Sky King Ranch   
3                            The Paris Apartment               -   
4                                       Dempsey         Tier One   
...                                          ...             ...   
6997                                 They Thirst               -   
6998                                        Snap               -   
6999                  The Traveling Vampire Show               -   
7000     WISHBONE...Be Careful What You Wish For               -   
7001                            Dropcloth Angels               -   

              Author_Name              Genre  Average_Rating  No_of_Ratings  \
0        J. Denison Reed    Action-Adventure            4.42             33   
1        Michael Crichton   Action-Adventure            4.09         944442   
2       Susan May Warren    Action-Adventure            4.43           1467   
3             Lucy Foley    Action-Adventure            3.67         473386   
4         Brian  Andrews    Action-Adventure            4.58           1708   
...                   ...                ...             ...            ...   
6997      Robert McCammon  Horror/Paranormal            3.91          12688   
6998      Steven Bredice   Horror/Paranormal            3.40             43   
6999       Richard Laymon  Horror/Paranormal            3.83           8813   
7000     Brooklyn Hudson   Horror/Paranormal            3.73            764   
7001  Gerald D. Johnston   Horror/Paranormal            4.27             15   

      No_of_Reviews  Score  No_of_Votes Publish_Year  
0                16    398            4   2021-05-27  
1             19172  30359          307   1990-07-11  
2               335    191            2   2022-01-06  
3             40240    296            3   2022-02-22  
4                86   1260           13   2023-02-21  
...             ...    ...          ...          ...  
6997            564   3615           46   1981-01-01  
6998             11    200            2   2012-01-01  
6999            645   3577           40   2000-01-04  
7000             94    200            2   2012-02-20  
7001              4    200            2   2012-10-29  

[7002 rows x 10 columns]

In [25]:
novel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7002 entries, 0 to 7001
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Novel_Name      7002 non-null   object        
 1   Series_Name     7002 non-null   object        
 2   Author_Name     7002 non-null   object        
 3   Genre           7002 non-null   object        
 4   Average_Rating  7002 non-null   float64       
 5   No_of_Ratings   7002 non-null   int32         
 6   No_of_Reviews   7002 non-null   int32         
 7   Score           7002 non-null   int32         
 8   No_of_Votes     7002 non-null   int32         
 9   Publish_Year    6988 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int32(4), object(4)
memory usage: 437.8+ KB


There are 6988 datas in `Publish_Year` so we drop all the rows which has null values in `Publish_Year`

In [26]:
novel.dropna(axis=0,inplace=True)

In [28]:
novel.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6988 entries, 0 to 7001
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Novel_Name      6988 non-null   object        
 1   Series_Name     6988 non-null   object        
 2   Author_Name     6988 non-null   object        
 3   Genre           6988 non-null   object        
 4   Average_Rating  6988 non-null   float64       
 5   No_of_Ratings   6988 non-null   int32         
 6   No_of_Reviews   6988 non-null   int32         
 7   Score           6988 non-null   int32         
 8   No_of_Votes     6988 non-null   int32         
 9   Publish_Year    6988 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int32(4), object(4)
memory usage: 491.3+ KB


In [29]:
#Creating a new column which contains only the year of Publish
novel["Publish_Year"]=pd.to_datetime(novel["Publish_Year"])
novel["Publish_year"]=novel["Publish_Year"].dt.year
novel.rename(columns={'Publish_Year':'Publish_Date'},inplace=True)

In [30]:
# Dropping the 'Score' column since it didn't have useful information
del novel["Score"]

In [31]:
novel.reset_index(drop=True,inplace=True)

In [32]:
novel

Novel_Name     Series_Name  \
0     Clifford's War: The Bluegrass Battleground               -   
1                                 Jurassic Park    Jurassic Park   
2                                      Sunburst   Sky King Ranch   
3                            The Paris Apartment               -   
4                                       Dempsey         Tier One   
...                                          ...             ...   
6983                                 They Thirst               -   
6984                                        Snap               -   
6985                  The Traveling Vampire Show               -   
6986     WISHBONE...Be Careful What You Wish For               -   
6987                            Dropcloth Angels               -   

              Author_Name              Genre  Average_Rating  No_of_Ratings  \
0        J. Denison Reed    Action-Adventure            4.42             33   
1        Michael Crichton   Action-Adventure            4.09         944442   
2       Susan May Warren    Action-Adventure            4.43           1467   
3             Lucy Foley    Action-Adventure            3.67         473386   
4         Brian  Andrews    Action-Adventure            4.58           1708   
...                   ...                ...             ...            ...   
6983      Robert McCammon  Horror/Paranormal            3.91          12688   
6984      Steven Bredice   Horror/Paranormal            3.40             43   
6985       Richard Laymon  Horror/Paranormal            3.83           8813   
6986     Brooklyn Hudson   Horror/Paranormal            3.73            764   
6987  Gerald D. Johnston   Horror/Paranormal            4.27             15   

      No_of_Reviews  No_of_Votes Publish_Date  Publish_year  
0                16            4   2021-05-27          2021  
1             19172          307   1990-07-11          1990  
2               335            2   2022-01-06          2022  
3             40240            3   2022-02-22          2022  
4                86           13   2023-02-21          2023  
...             ...          ...          ...           ...  
6983            564           46   1981-01-01          1981  
6984             11            2   2012-01-01          2012  
6985            645           40   2000-01-04          2000  
6986             94            2   2012-02-20          2012  
6987              4            2   2012-10-29          2012  

[6988 rows x 10 columns]

In [33]:
novel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6988 entries, 0 to 6987
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Novel_Name      6988 non-null   object        
 1   Series_Name     6988 non-null   object        
 2   Author_Name     6988 non-null   object        
 3   Genre           6988 non-null   object        
 4   Average_Rating  6988 non-null   float64       
 5   No_of_Ratings   6988 non-null   int32         
 6   No_of_Reviews   6988 non-null   int32         
 7   No_of_Votes     6988 non-null   int32         
 8   Publish_Date    6988 non-null   datetime64[ns]
 9   Publish_year    6988 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int32(3), int64(1), object(4)
memory usage: 464.2+ KB


In [34]:
cleaned_novel=novel.copy()

In [36]:
#saving the cleaned DataFrame as csv file
cleaned_novel=cleaned_novel.to_csv(r"Data Sets\cleaned_novel_dataset.csv",index=False)